src/parser/toplevel.lisp

```diff
@@ -554,56 +536,213 @@ consume all attributes"))))
-                 :primary-note "missing package name")))
+           (values toplevel-package))
+  (let ((iter (cons form file)))
+    (require-symbol iter "PACKAGE" "package declarations must start with `package`")
+    (let* ((package-name (cst:raw (consume-symbol iter :missing "missing package name"
+                                                       :not-symbol "package name must be a symbol" )))
+           (package (make-instance 'toplevel-package
+                      :name (symbol-name package-name)
+                      :docstring (consume-string iter :optional t))))
+      (loop :while (next-value-p iter)
+            :do (parse-package-clause package (consume-value iter) file))
+      package)))
 
-  ;; Package declarations cannot contain more than two forms
-  (when (cst:consp (cst:rest (cst:rest form)))
-    (error 'parse-error
-           :err (se:source-error
-                 :span (cst:source (cst:first (cst:rest (cst:rest form))))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "unexpected forms")))
-
-  (unless (identifierp (cst:raw (cst:second form)))
-    (error 'parse-error
-           :err (se:source-error
-                 :span (cst:source (cst:second form))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "package name must be a symbol")))
-
-  (let* ((package-name (symbol-name (cst:raw (cst:second form))))
-
-         (package (find-package package-name)))
-
-    (unless package
-      (setf package (make-package package-name :use '("COALTON" "COALTON-PRELUDE"))))
+;; Empty package for reading (package) forms
 
-    package))
+(defpackage #:coalton-impl/parser/%defpackage
+  (:documentation "The sole purpose of this package is to provide storage for symbols that are interned while Coalton package forms are read.")
+  (:use))
 
+(defun read-toplevel-package (stream file)
+  "Read and parse a Coalton toplevel package form."
+  (let ((*package* (find-package 'coalton-impl/parser/%defpackage)))
+    (unwind-protect
+         (multiple-value-bind (form presentp)
+             (maybe-read-form stream *coalton-eclector-client*)
+           (unless presentp
+             (error 'parse-error
+                    :err (se:source-error
+                          :span (cons (- (file-position stream) 2)
+                                      (- (file-position stream) 1))
+                          :file file
+                          :message "Unexpected EOF"
+                          :primary-note "missing package form")))
+           (parse-package form file))
+      (do-symbols (symbol)
+        (unintern symbol)))))
+
+(defun make-defpackage (toplevel-package)
+  "Generate a Lisp defpackage form from a Caolton toplevel-package structure."
+  (with-slots (name import import-from import-as export) toplevel-package
+    `(defpackage ,name
```

Comment:
uiop:define-package

Reply:
using eval form for now - there will be a subsequent MR with missing package forms ('shadow') that would be a good place to make semantics of (package) final